pkg/service/secret.go

```diff
@@ -0,0 +1,98 @@
+package service
+
+import (
+	"github.com/appuio/seiso/cfg"
+	"github.com/appuio/seiso/pkg/kubernetes"
+	"github.com/appuio/seiso/pkg/openshift"
+	log "github.com/sirupsen/logrus"
+	"github.com/thoas/go-funk"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	core "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+type SecretServiceInt interface {
+	ListNamesAndLabels() (resourceNames, labels []string, err error)
+	List(listOptions metav1.ListOptions) (resources []cfg.KubernetesResource, err error)
+	GetUnused(namespace string, resources []cfg.KubernetesResource) (unusedResources []cfg.KubernetesResource, funcErr error)
+	Delete(resources []cfg.KubernetesResource)
+}
+
+type SecretService struct {
+	Secret core.SecretInterface
+	Helper kubernetes.Kubernetes
+}
+
+func (ss SecretService) ListNamesAndLabels() (resourceNames, labels []string, err error) {
+	secrets, err := ss.List(metav1.ListOptions{})
+	secretNames, labels := GetNamesAndLabels(secrets)
+	return secretNames, labels, nil
+}
+
+// ListSecrets returns a list of secrets from a namespace
+func (ss SecretService) List(listOptions metav1.ListOptions) (resources []cfg.KubernetesResource, err error) {
+	secrets, err := ss.Secret.List(listOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, secret := range secrets.Items {
+		resource := cfg.NewSecretResource(
+			secret.GetName(),
+			secret.GetNamespace(),
+			secret.GetCreationTimestamp().Time,
+			secret.GetLabels())
+		resources = append(resources, resource)
+	}
+
+	return resources, nil
+}
+
+// ListUnusedResources lists resources that are unused
+func (ss SecretService) GetUnused(namespace string, resources []cfg.KubernetesResource) (unusedResources []cfg.KubernetesResource, funcErr error) {
+	var usedResources []cfg.KubernetesResource
+	funk.ForEach(openshift.PredefinedResources, func(predefinedResource schema.GroupVersionResource) {
+		funk.ForEach(resources, func(resource cfg.KubernetesResource) {
+
+			resourceName := resource.GetName()
+
+			if funk.Contains(usedResources, resource) {
+				// already marked as existing, skip this
+				return
+			}
+			contains, err := ss.Helper.ResourceContains(namespace, resourceName, predefinedResource)
+			if err != nil {
+				funcErr = err
+				return
+			}
+
+			if contains {
+				usedResources = append(usedResources, resource)
+			}
+		})
+	})
+
+	for _, resource := range resources {
+		if !funk.Contains(usedResources, resource) {
+			unusedResources = append(unusedResources, resource)
+		}
+	}
+
+	return unusedResources, funcErr
+}
+
+func (ss SecretService) Delete(resources []cfg.KubernetesResource) {
```

Comment:
This should expect `[]v1.Secrets`, not interfaces

Reply:
Do we want to duplicate the code or to create a common service with ConfigMaps?

Reply:
Actually we perform some common operations, so it's not a bad idea to use a common structure